core/distributed/matrix.cpp

```diff
@@ -252,14 +258,105 @@ void Matrix<ValueType, LocalIndexType, GlobalIndexType>::read_distributed(
+                        overlap_recv_sizes.data(), 1);
+        std::partial_sum(overlap_recv_sizes.begin(), overlap_recv_sizes.end(),
+                         overlap_recv_offsets.begin() + 1);
+        overlap_send_offsets[0] = 0;
+        overlap_recv_offsets[0] = 0;
+
+        size_type n_send = overlap_send_offsets.back();
+        size_type n_recv = overlap_recv_offsets.back();
+        array<global_index_type> overlap_send_row_idxs{exec, n_send};
+        array<global_index_type> overlap_send_col_idxs{exec, n_send};
+        array<value_type> overlap_send_values{exec, n_send};
+        array<global_index_type> overlap_recv_row_idxs{exec, n_recv};
+        array<global_index_type> overlap_recv_col_idxs{exec, n_recv};
+        array<value_type> overlap_recv_values{exec, n_recv};
+        exec->run(matrix::make_fill_overlap_send_buffers(
+            data, tmp_part.get(), local_part, overlap_positions,
+            original_positions, overlap_send_row_idxs, overlap_send_col_idxs,
+            overlap_send_values));
+
+        if (use_host_buffer) {
+            overlap_send_row_idxs.set_executor(exec->get_master());
+            overlap_send_col_idxs.set_executor(exec->get_master());
+            overlap_send_values.set_executor(exec->get_master());
+            overlap_recv_row_idxs.set_executor(exec->get_master());
+            overlap_recv_col_idxs.set_executor(exec->get_master());
+            overlap_recv_values.set_executor(exec->get_master());
+        }
+        comm.all_to_all_v(
+            use_host_buffer ? exec : exec->get_master(),
+            overlap_send_row_idxs.get_const_data(), overlap_send_sizes.data(),
+            overlap_send_offsets.data(), overlap_recv_row_idxs.get_data(),
+            overlap_recv_sizes.data(), overlap_recv_offsets.data());
+        comm.all_to_all_v(
+            use_host_buffer ? exec : exec->get_master(),
+            overlap_send_col_idxs.get_const_data(), overlap_send_sizes.data(),
+            overlap_send_offsets.data(), overlap_recv_col_idxs.get_data(),
+            overlap_recv_sizes.data(), overlap_recv_offsets.data());
+        comm.all_to_all_v(
+            use_host_buffer ? exec : exec->get_master(),
+            overlap_send_values.get_const_data(), overlap_send_sizes.data(),
+            overlap_send_offsets.data(), overlap_recv_values.get_data(),
+            overlap_recv_sizes.data(), overlap_recv_offsets.data());
+        if (use_host_buffer) {
+            overlap_recv_row_idxs.set_executor(exec);
+            overlap_recv_col_idxs.set_executor(exec);
+            overlap_recv_values.set_executor(exec);
+        }
+
+        array<global_index_type> all_row_idxs{exec, num_entries + n_recv};
+        array<global_index_type> all_col_idxs{exec, num_entries + n_recv};
+        array<value_type> all_values{exec, num_entries + n_recv};
+        exec->copy_from(exec, num_entries, data.get_const_row_idxs(),
+                        all_row_idxs.get_data());
```

Comment:
For another time, but I think this would be clearer if we could do:
```
all_row_idxs.subview(0, num_entries) = data.row_idxs;
all_row_idxs.subview(num_entries, all_size) = overlap_recv_row_idxs;
```
But this requires to expose array acces in `device_matrix_data`.